src/darwin/Framework/CHIPTests/MTRCommissionableBrowserTests.m

```diff
@@ -0,0 +1,266 @@
+    XCTAssertFalse([controller isRunning]);
+
+    [[MTRDeviceControllerFactory sharedInstance] stopControllerFactory];
+}
+
+- (void)testStartBrowseAndStopBrowse
+{
+    // Check that there is no ongoing browse
+    XCTAssertFalse([sController isBrowsing]);
+
+    __auto_type delegate = [[DeviceScannerDelegate alloc] init];
+    dispatch_queue_t dispatchQueue = dispatch_queue_create("com.chip.discover", DISPATCH_QUEUE_SERIAL);
+
+    // Start browsing
+    XCTAssertTrue([sController startBrowse:delegate queue:dispatchQueue]);
+    XCTAssertTrue([sController isBrowsing]);
+
+    // Stop browsing
+    XCTAssertTrue([sController stopBrowse]);
+    XCTAssertFalse([sController isBrowsing]);
+}
+
+- (void)testStartBrowseAndStopBrowseMultipleTimes
+{
+    // Check that there is no ongoing browse
+    XCTAssertFalse([sController isBrowsing]);
+
+    __auto_type delegate = [[DeviceScannerDelegate alloc] init];
+    dispatch_queue_t dispatchQueue = dispatch_queue_create("com.chip.discover", DISPATCH_QUEUE_SERIAL);
+
+    // Start browsing
+    XCTAssertTrue([sController startBrowse:delegate queue:dispatchQueue]);
+    XCTAssertTrue([sController isBrowsing]);
+
+    // Stop browsing
+    XCTAssertTrue([sController stopBrowse]);
+    XCTAssertFalse([sController isBrowsing]);
+
+    // Start browsing a second time
+    XCTAssertTrue([sController startBrowse:delegate queue:dispatchQueue]);
+    XCTAssertTrue([sController isBrowsing]);
+
+    // Stop browsing a second time
+    XCTAssertTrue([sController stopBrowse]);
+    XCTAssertFalse([sController isBrowsing]);
+}
+
+- (void)testStopBrowseWhileNotBrowsing
+{
+    // Check that there is no ongoing browse
+    XCTAssertFalse([sController isBrowsing]);
+
+    // Stop browsing while there is no browse ongoing
+    XCTAssertFalse([sController stopBrowse]);
+    XCTAssertFalse([sController isBrowsing]);
+}
+
+- (void)testStartBrowseWhileBrowsing
+{
+    // Check that there is no ongoing browse
+    XCTAssertFalse([sController isBrowsing]);
+
+    __auto_type delegate = [[DeviceScannerDelegate alloc] init];
+    dispatch_queue_t dispatchQueue = dispatch_queue_create("com.chip.discover", DISPATCH_QUEUE_SERIAL);
+
+    // Start browsing
+    XCTAssertTrue([sController startBrowse:delegate queue:dispatchQueue]);
+    XCTAssertTrue([sController isBrowsing]);
+
+    // Start browsing a second time while a browse is ongoing
+    XCTAssertFalse([sController startBrowse:delegate queue:dispatchQueue]);
+    XCTAssertTrue([sController isBrowsing]);
+
+    // Properly stop browsing
+    XCTAssertTrue([sController stopBrowse]);
+    XCTAssertFalse([sController isBrowsing]);
+}
+
+- (void)testStartBrowseGetCommissionableOverMdns
+{
+    // Check that there is no ongoing browse
+    XCTAssertFalse([sController isBrowsing]);
+
+    __auto_type expectation = [self expectationWithDescription:@"Commissionable Device Found"];
+    __auto_type delegate = [[DeviceScannerDelegate alloc] initWithExpectation:expectation];
```

Comment:
Feels like here we should also init it with some sort of "set of things to expect you will discover"...

Reply:
I made that sort of static for now. It could be revisited if needed.